General reversible binary op support (e.g. __add__ / __radd__) in dynamo

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93271

Generic support for reversible binary op pairs (e.g. `__add__` / `__radd__`) in dynamo.
Adds logic to flip args and try the reverse op when the forward op is unsupported.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

jbschlosser committed:
Support non-symbolic - symbolic subtraction via __rsub__

[ghstack-poisoned]

jbschlosser committed:
Update on "Support non-symbolic - symbolic subtraction via __rsub__"


Uses `__rsub__` to support things like `5 - x.shape[0]`.

Note:
* `5 - x.shape[0]` takes the `ConstantVariable - DynamicShapeVariable` codepath
* `param.shape[0] - x.shape[0]` takes the `DynamicShapeVariable - DynamicShapeVariable` codepath

In general, it seems strange that we are making the assumption that parameter shapes are static, but they still end up tracked by DynamicShapeVariables instead of ConstantVariables. Should we change this?

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -551,6 +551,10 @@ def call_add(self, tx, *args, **kwargs):
     def call_sub(self, tx, *args, **kwargs):
+        # Commute sub when the right-hand side is symbolic
+        # (since e.g. int.__sub__(int, SymInt) is not implemented).
+        if len(args) == 2 and args[1].python_type() in (torch.SymInt, torch.SymFloat):
+            return args[1].call_method(tx, "__rsub__", [args[0]], kwargs)
```
If you convert arg[0] into DynamicShapeVariable, you can avoid ever hitting `if name == "__sub__":` in ConstantVariable blow, and just fall into `def call_method(` on DynamicShapeVariable, right?

ezyang replied:
You should figure out how to uniformly apply this fix to all of the reversible call_ methods

voznesenskym reviewed (changes requested):
Actually there may be a simpler impl?

ezyang commented on torch/_dynamo/variables/constant.py:
```diff
@@ -98,6 +98,10 @@ def call_method(
+            # For non-commutative op sub, swap args and invoke rsub.
+            if name == "__sub__":
+                assert len(args) == 1
+                return args[0].call_method(tx, "__rsub__", [self], {})
```
fwiw, we should use radd for add as well, because that more closely preserves original semantics

ezyang replied:
Also, there is a clear opportunity to correctly implement this for *all* reversible magic methods, not just sub.

jbschlosser commented:
Thanks for the feedback @voznesenskym / @ezyang  - reworking this to be more general.

jbschlosser committed:
Update on "Support non-symbolic - symbolic subtraction via __rsub__"


Uses `__rsub__` to support things like `5 - x.shape[0]`.

Note:
* `5 - x.shape[0]` takes the `ConstantVariable - DynamicShapeVariable` codepath
* `param.shape[0] - x.shape[0]` takes the `DynamicShapeVariable - DynamicShapeVariable` codepath. Since DynamicShapeVariables can contain constants, we need to handle commuting logic for this case as well

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -340,6 +373,15 @@ def call_function(
+        # Call reverse (e.g. __radd__) if forward (e.g. __add__) isn't supported.
+        # This can be the case when doing int + SymInt, for example.
+        forward_res = args[0].call_method(tx, forward_name, args[1:], kwargs)
+        if forward_res is NotImplemented:
```
This is a little suspect, because the method should return a variable tracker, not a "real" NotImplemented

ezyang replied:
You should return a ConstantVariable of NotImplemented. You also need to test if the rhs is NotImplemented and raise an error (probably just unimplemented?) if so, since both sides missing induces TypeError (but Dynamo's general handling of error cases is to bail)

ezyang reviewed (commented):
This looks really good! Just some small final adjustments. Also giving Voz a chance to comment

voznesenskym commented on torch/_dynamo/testing.py:
```diff
@@ -215,7 +221,7 @@ def standard_test(self, fn, nargs, expected_ops=None, expected_ops_dynamic=None)
     self.assertTrue(same(val2a, correct2))
     self.assertTrue(same(val2b, correct2))
-    self.assertEqual(actual.frame_count, 1)
+    self.assertEqual(actual.frame_count, expected_frame_count)
```
No, let's keep this at 1 - this test explicitly uses `optimize_assert` for that end - if you want frame counts of > 1, check out usage of compile counter.

I would be open to a version of this fn that calls `torch._dynamo.optimize(compile_counter)` - search for "torch._dynamo.testing.CompileCounter()" in the suite for examples.

jbschlosser replied:
Ah I see - `optimize_assert` ensures no graph breaks. I reverted this change and tweaked my tests slightly to have a single operation instead of being constant folded out when dynamic shapes are disabled.

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -942,6 +942,12 @@ def _clone_input(value):
     elif isinstance(example_value, (torch.SymInt, torch.SymFloat)):
         proxy.node.meta["example_value"] = example_value
         return DynamicShapeVariable(proxy, example_value, **options)
+    elif example_value is NotImplemented:
```
Riding NotImplemented in example_value is weird, to me - where is it coming from?

jbschlosser replied:
The naming of `example_value` was a bit weird to me, but it comes from:
```
example_value = get_fake_value(proxy.node, tx)
```

so it's the result of running the node computation using fake tensors. `get_fake_value()` will return `NotImplemented` for the cases we're interested in.

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -942,6 +942,12 @@ def _clone_input(value):
+        # We want the NotImplemented to propagate upward.
+        # For example, if __add__ returns NotImplemented, we should try __radd__
+        # and leave __add__ out of the graph.
+        proxy.node.graph.erase_node(proxy.node)
```
See above, I think this flow is sus

voznesenskym commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -340,6 +373,15 @@ def call_function(
+    def _call_reversible(self, tx, forward_name, reverse_name, args, kwargs):
+        # Call reverse (e.g. __radd__) if forward (e.g. __add__) isn't supported.
+        # This can be the case when doing int + SymInt, for example.
+        forward_res = args[0].call_method(tx, forward_name, args[1:], kwargs)
+        if forward_res is NotImplemented:
```
If you see a NotImplemented here, you need to do the same probe for `args[1].call_method(tx, reverse_name, [args[0]], kwargs)` - and if that returns a NotImplemented, you should just invoke unimplemented.

This fn should return either a VariableBuilder, or raise unimplemented as @ezyang said.

voznesenskym commented on torch/_dynamo/variables/constant.py:
```diff
@@ -87,19 +87,11 @@ def call_method(
-            # Unfortunate constant
-            return super(ConstantVariable, self).call_method(tx, name, args, kwargs)
+            # Promote to DynamicShapeVariable for operations involving dynamic shapes.
+            return variables.DynamicShapeVariable(
+                self.as_proxy(), self.value
+            ).call_method(tx, name, args, kwargs)
```
Ah, you hit a gap in our implementation! So this is a little unsupported - ConstantVariable has no real proxy  (self.as_proxy() on a constant just gives us the real value) - meanwhile, DynamicShapeVariable has assumptions around proxy being both real and there - I would instead suggest either:

1) making a new proxy from scratch to track operations, with a ConstantSource - there is some unfortunate code in register_attr_or_module that might serve as reference

2) Make a dead-end variable - something like NumericalVariable - that is proxy-less, like ConstantSource, but knows how to do numerical operations on Dyn and Consts together (worse option)

cc: @ezyang  for input too

ezyang replied:
What's wrong with the proxy just being the Python literal? You can still trace with it (as long as there's a proxy somewhere, which there is)

jbschlosser committed:
Update on "Reversible binary op support (e.g. __add__ / __radd__)"


Support for reversable binary ops in dynamo.

Caveats:
* Node in FX graph isn't nice yet :( e.g. it'll be `__add__` instead of `operator.add`

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser committed:
Update on "General reversible binary op support (e.g. __add__ / __radd__) in dynamo"


Generic support for reversible binary op pairs (e.g. `__add__` / `__radd__`) in dynamo.
Adds logic to flip args and try the reverse op when the forward op is unsupported.

Caveats:
* Node in FX graph isn't nicely named :( (e.g. it'll show up as `__add__` instead of `operator.add`)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -306,6 +331,12 @@ def call_function(
             )
             return out
 
```
Idea: Put a carve out for ConstantVariable-DynamicShapeVariable here

ezyang replied:
we basically want to do

```

        options = VariableTracker.propagate(self, args, kwargs.values())

        return wrap_fx_proxy(
            tx,
            tx.output.create_proxy(
                "call_function",
                self.fn,
                *proxy_args_kwargs([self] + list(args), kwargs),
            ),
            **options,
        )

```

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -306,6 +331,12 @@ def call_function(
+        reversible_functions = self._reversible_functions()
+        if self.fn in reversible_functions:
+            forward_name, reverse_name = reversible_functions[self.fn]
+            return self._call_reversible(tx, forward_name, reverse_name, args, kwargs)
```
still want to keep this for UserDefined

jbschlosser committed:
Update on "General reversible binary op support (e.g. __add__ / __radd__) in dynamo"


Generic support for reversible binary op pairs (e.g. `__add__` / `__radd__`) in dynamo.
Adds logic to flip args and try the reverse op when the forward op is unsupported.

Caveats:
* Node in FX graph isn't nicely named :( (e.g. it'll show up as `__add__` instead of `operator.add`)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -306,6 +332,63 @@ def call_function(
+                return self._dyn_proxy(tx, *args, **kwargs)
+
+            # Handle special cases for list / tuple.
+            # TODO: Develop a proper abstraction to avoid this nonsense
```
This is the part I'm talking about - specific customizations for lists / tuples that can no longer be handled through the `get_method()` abstraction

jansel reviewed (changes requested):
Overall I like this approach, though it would be good to search the dynamo codebase for the methods that are being removed.  There are a few cases like:
https://github.com/pytorch/pytorch/blob/10910758f462a7b673b60c4755c553ba2abb462d/torch/_dynamo/variables/lists.py#L87-L89
that we should remove since they are now unreachable.

After that I'd suggest adding some logic to the base version of call_method():
https://github.com/pytorch/pytorch/blob/10910758f462a7b673b60c4755c553ba2abb462d/torch/_dynamo/variables/base.py#L232
that does stuff like:
```
if name == "__add__":
  return BuiltinVariable(operator.add).call_method(tx, [self] + args, kwargs)
```

This will handle the case of the user writing code like:
```
[1].__add__([2])
```

ezyang commented:
> After that I'd suggest adding some logic to the base version of call_method():

Don't do this, `x.__add__(y)` is different than `x + y` ;)

```
>>> 1 + object()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for +: 'int' and 'object'
>>> (1).__add__(object())
NotImplemented
```

jbschlosser committed:
Update on "General reversible binary op support (e.g. __add__ / __radd__) in dynamo"


Generic support for reversible binary op pairs (e.g. `__add__` / `__radd__`) in dynamo.
Adds logic to flip args and try the reverse op when the forward op is unsupported.

Caveats:
* Gnarly set of isinstance checks in `BuiltinVariable.call_function()` with list / tuple specific logic

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -162,6 +162,122 @@ def _reversible_binops():
+        ]
+        op_handlers[operator.mul].extend(list_like_expansion_handlers)
+
+        return op_handlers
```
This seems fine.

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -162,6 +162,122 @@ def _reversible_binops():
+                if isinstance(a, UserDefinedVariable):
+                    return a.call_method(tx, forward_name, [b], {})
+                else:
+                    return b.call_method(tx, reverse_name, [a], {})
```
This is technically incorrect because when you have VariableTracker + UserDefinedVariable, whether or not you end up going to `__radd__` depends on the specifics of VariableTracker. For something like int, it will always go to the user defined variable. But here is a naughty example:

```
import torch

class A(int):
    def __radd__(self, other):
        print("woof")

torch.randn(3) + A(3)
```

this prints nothing, because Tensor.__add__ only does a subtype test against int and will ignore the subclass. (Arguably Tensor is wrong in doing it this way, but c'est la vie.)

jbschlosser replied:
Oh interesting! Nice example, and it sadly invalidates my assumption that this is okay to do. It sounds like we can't get away from actually running the logic and checking for `NotImplemented` -> remove the node from the graph?

ezyang replied:
Well, not quite. In this case, you need to know something about how `Tensor.__add__` is implemented. Now that you know that it never returns NotImplemented if the RHS is an int subtype, you could hard code that here.

jbschlosser replied:
@ezyang should I do something about this before landing or just drop in a TODO?

jbschlosser replied:
@ezyang Realized your test case will not hit the logic in this PR, since it's limited to binary op invocation with *non-tensor* args. I think there's a place for expanding it to cover tensor args and other built-in op logic (e.g. helping #92618 expand dynamic shape support, etc.), but probably in follow-up PRs. wdyt?

ezyang replied:
Oh yeah, sorry I wasn't clear, I approved because this is OK to TODO. My point is also more general: to be super pedantically correct, we need to reason this out for each case. But it's unlikely to matter IRL

ezyang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -162,6 +162,122 @@ def _reversible_binops():
+            ),
+            (
+                (variables.ConstantVariable, variables.TupleVariable),
+                lambda tx, a, b, options: expand_list_like(tx, b, a, options),
```
I didn't review the list stuff carefully, assuming it's similar to what we had before

jbschlosser commented:
> Overall I like this approach, though it would be good to search the dynamo codebase for the methods that are being removed.

FYI I did this already :)

jbschlosser committed:
Update on "General reversible binary op support (e.g. __add__ / __radd__) in dynamo"


Generic support for reversible binary op pairs (e.g. `__add__` / `__radd__`) in dynamo.
Adds logic to flip args and try the reverse op when the forward op is unsupported.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser committed:
Update on "General reversible binary op support (e.g. __add__ / __radd__) in dynamo"


Generic support for reversible binary op pairs (e.g. `__add__` / `__radd__`) in dynamo.
Adds logic to flip args and try the reverse op when the forward op is unsupported.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
